components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImpl.java

```diff
@@ -0,0 +1,246 @@
+    {
+        return new Function<Object, Iterable<PrimaryEntity>>()
+        {
+            @Override
+            public Iterable<PrimaryEntity> apply(final Object eid)
+            {
+                try {
+                    final Query q = qm.createQuery("where doc.object(PhenoTips.PatientClass)"
+                        + ".external_id = :eid", Query.XWQL);
+                    q.bindValue("eid", eid);
+                    final List<String> patientIds = q.execute();
+                    return getPatientsFromIds(patientIds);
+                } catch (final QueryException ex) {
+                    logger.warn("Failed to retrieve patient with external id [{}]: {}", eid, ex.getMessage());
+                }
+                return Collections.emptyList();
+            }
+        };
+    }
+
+    /**
+     * Gets patient entities, given a list of patient IDs. Any patient IDs that are not in the database will
+     * be ignored.
+     *
+     * @param patientIds the list of patient IDs of interest -- should not be null
+     * @return an iterable containing patient entity objects
+     */
+    @Nonnull
+    private Iterable<PrimaryEntity> getPatientsFromIds(@Nonnull final List<String> patientIds)
+    {
+        return FluentIterable.from(patientIds).transform(this.patientFromIdFunction)
+            .filter(Predicates.<PrimaryEntity>notNull());
+    }
+
+    /**
+     * Returns the patient entity if it exists and if the user has view rights. Otherwise returns null.
+     *
+     * @param patientId the ID of the patient of interest
+     * @return the patient entity with the specified ID, if exists, null otherwise
+     */
+    @Nullable
+    private PrimaryEntity getPatientFromId(final String patientId)
+    {
+        try {
+            final PrimaryEntity patient = repository.get(patientId);
+            final User currentUser = this.users.getCurrentUser();
+            // If user has view rights and patient with the provided id exists, return the patient. Else return null.
+            if (patient != null && this.access.hasAccess(Right.VIEW, currentUser == null ? null
+                    : currentUser.getProfileDocument(), patient.getDocument())) {
+                return patient;
+            }
+            this.logger.debug("Patient not found, or view access denied to user [{}] on patient record [{}]",
+                currentUser, patientId);
+        } catch (final IllegalArgumentException ex) {
+            logger.warn("Failed to retrieve patient with ID [{}]: {}", patientId, ex.getMessage());
+        }
+        return null;
+    }
+
+    /**
+     * Gets the function that gets the patient given the patient's ID.
+     *
+     * @return a function that gets the patient associated with the provided ID
+     */
+    @Nonnull
+    private Function<String, PrimaryEntity> getPatientFromIdFunction()
+    {
+        return new Function<String, PrimaryEntity>()
```

Comment:
Regarding Guava, it's already included in the project. In my opinion, switching to loops, in this case, would make the code more bulky, and could potentially introduce new bugs.